File: frontend/src/lib/likely-locations.ts
import type { LikelyLocation } from "@/types/incident";
import type { GeoPoint } from "@/types/telemetry";

const DEFAULT_MISSING_MINUTES = 60;
const ON_TRAIL_PACE_MPS = 1.1;
const INJURED_PACE_MPS = 0.7;
const MEDIAN_DISTANCE_M = 700;
const SPREAD_M = 500;
const OFF_TRAIL_STEP_M = 40;
const MIN_SEPARATION_M = 80;
const TOP_N = 5;
const TURN_DEGREES = 35;
const EARTH_M = 6_371_000;

export function scoreLikelyLocations(
  trail: GeoPoint[],
  pls: GeoPoint | null,
  missingMinutes: number,
  transcript: string,
): LikelyLocation[] {
  if (trail.length < 2) {
    return [];
  }
  const minutes = Math.max(1, missingMinutes);
  const lowered = transcript.toLowerCase();
  const injured = ["ankle", "injur", "immobile", "sprain"].some((cue) => lowered.includes(cue));
  const pace = injured ? INJURED_PACE_MPS : ON_TRAIL_PACE_MPS;
  const maxMeters = minutes * 60 * pace;
  const along = alongTrailMeters(trail);
  const snap = pls ? nearestIndex(trail, pls) : 0;
  const plsAlong = along[snap] ?? 0;
  const leftTrail = ["left the trail", "off the trail", "off the main trail", "left the path"].some(
    (cue) => lowered.includes(cue),
  );

  const candidates: LikelyLocation[] = [];
  trail.forEach((point, index) => {
    const distance = Math.abs((along[index] ?? 0) - plsAlong);
    if (distance > maxMeters) {
      return;
    }
    const extras: string[] = [];
    let bonus = 0;
    if (isJunction(trail, index)) {
      bonus += 0.15;
      extras.push("switchback / decision point");
    }
    if (index === 0 || index === trail.length - 1) {
      bonus += 0.08;
      extras.push("trail end");
    }
    if (["switchback"].some((cue) => lowered.includes(cue)) && isJunction(trail, index)) {
      bonus += 0.2;
      extras.push("switchback mentioned");
    }
    if (
      ["viewpoint", "lookout", "overlook"].some((cue) => lowered.includes(cue)) &&
      index >= Math.floor(trail.length * 0.55)
    ) {
      bonus += 0.2;
      extras.push("viewpoint cue");
    }
    if (
      [" the y", "past the y", "below the y", "the y itself", "y mountain"].some((cue) =>
        lowered.includes(cue),
      ) &&
      index >= Math.floor(trail.length * 0.7)
    ) {
      bonus += 0.2;
      extras.push("Y landmark");
    }
    if (["uphill", "up the", "went up"].some((cue) => lowered.includes(cue)) && (along[index] ?? 0) > plsAlong) {
      bonus += 0.1;
      extras.push("uphill of PLS");
    }
    const decay = Math.exp(-0.5 * ((distance - MEDIAN_DISTANCE_M) / SPREAD_M) ** 2);
    const raw = decay + bonus;
    const reason = reasonFor(distance, extras);
    candidates.push({
      point,
      score: raw,
      reason,
      distanceFromPlsMeters: distance,
    });
    if (isJunction(trail, index)) {
      candidates.push({
        point: offsetPerpendicular(trail, index, OFF_TRAIL_STEP_M),
        score: raw + (leftTrail ? 0.12 : 0.06),
        reason: `${reason}; possible leave at decision point`,
        distanceFromPlsMeters: distance,
      });
    }
  });

  const kept = suppressNearDuplicates(candidates);
  const peak = Math.max(...kept.map((location) => location.score), 0);
  if (peak <= 0) {
    return [];
  }
  return kept
    .map((location) => ({
      ...location,
      score: Number((location.score / peak).toFixed(3)),
      distanceFromPlsMeters: Number(location.distanceFromPlsMeters.toFixed(1)),
    }))
    .sort((a, b) => b.score - a.score)
    .slice(0, TOP_N);
}

export function resolvedMissingMinutes(stored: number | null | undefined): {
  minutes: number;
  assumed: boolean;
} {
  if (stored == null) {
    return { minutes: DEFAULT_MISSING_MINUTES, assumed: true };
  }
  return { minutes: stored, assumed: false };
}

function reasonFor(distance: number, extras: string[]): string {
  const base = `Along-trail, ${(distance / 1000).toFixed(1)} km from PLS`;
  return extras.length > 0 ? `${base} — ${extras.join(", ")}` : `${base} — time-reachable`;
}

function alongTrailMeters(trail: GeoPoint[]): number[] {
  const distances = [0];
  for (let index = 1; index < trail.length; index += 1) {
    distances.push(distances[index - 1] + haversineMeters(trail[index - 1], trail[index]));
  }
  return distances;
}

function nearestIndex(trail: GeoPoint[], point: GeoPoint): number {
  let best = 0;
  let bestDistance = Number.POSITIVE_INFINITY;
  trail.forEach((candidate, index) => {
    const distance = haversineMeters(candidate, point);
    if (distance < bestDistance) {
      best = index;
      bestDistance = distance;
    }
  });
  return best;
}

function isJunction(trail: GeoPoint[], index: number): boolean {
  if (index <= 0 || index >= trail.length - 1) {
    return false;
  }
  const incoming = bearingDegrees(trail[index - 1], trail[index]);
  const outgoing = bearingDegrees(trail[index], trail[index + 1]);
  const delta = Math.abs(((outgoing - incoming + 180) % 360) - 180);
  return delta >= TURN_DEGREES;
}

function bearingDegrees(start: GeoPoint, end: GeoPoint): number {
  const lat1 = toRad(start.lat);
  const lat2 = toRad(end.lat);
  const dlng = toRad(end.lng - start.lng);
  const x = Math.sin(dlng) * Math.cos(lat2);
  const y = Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(dlng);
  return ((Math.atan2(x, y) * 180) / Math.PI + 360) % 360;
}

function offsetPerpendicular(trail: GeoPoint[], index: number, meters: number): GeoPoint {
  const start = index <= 0 ? trail[0] : trail[index - 1];
  const end = index <= 0 ? trail[1] : trail[index];
  const bearing = toRad((bearingDegrees(start, end) + 90) % 360);
  const north = meters * Math.cos(bearing);
  const east = meters * Math.sin(bearing);
  const origin = trail[index];
  return {
    lat: origin.lat + north / 111_320,
    lng: origin.lng + east / (111_320 * Math.cos(toRad(origin.lat))),
  };
}

function suppressNearDuplicates(candidates: LikelyLocation[]): LikelyLocation[] {
  const kept: LikelyLocation[] = [];
  [...candidates]
    .sort((a, b) => b.score - a.score)
    .forEach((location) => {
      if (kept.every((other) => haversineMeters(location.point, other.point) >= MIN_SEPARATION_M)) {
        kept.push(location);
      }
    });
  return kept;
}

function haversineMeters(a: GeoPoint, b: GeoPoint): number {
  const dlat = toRad(b.lat - a.lat);
  const dlng = toRad(b.lng - a.lng);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const h =
    Math.sin(dlat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dlng / 2) ** 2;
  return 2 * EARTH_M * Math.asin(Math.min(1, Math.sqrt(h)));
}

function toRad(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

File: backend/services/locate/likely_locations.py
from __future__ import annotations

import math
from collections.abc import Sequence

from models.domain import GeoPoint, LikelyLocation
from services.gis.route_metrics import haversine_m

DEFAULT_MISSING_MINUTES = 60
ON_TRAIL_PACE_MPS = 1.1
INJURED_PACE_MPS = 0.7
MEDIAN_DISTANCE_M = 700.0
SPREAD_M = 500.0
OFF_TRAIL_STEP_M = 40.0
MIN_SEPARATION_M = 80.0
TOP_N = 5
TURN_DEGREES = 35.0

_INJURY_CUES = ("ankle", "injur", "immobile", "sprain")
_LEFT_TRAIL_CUES = ("left the trail", "off the trail", "off the main trail", "left the path")
_SWITCHBACK_CUES = ("switchback",)
_VIEWPOINT_CUES = ("viewpoint", "lookout", "overlook")
_Y_CUES = (" the y", "past the y", "below the y", "the y itself", "y mountain")
_UPHILL_CUES = ("uphill", "up the", "went up")


def score_likely_locations(
    trail: Sequence[GeoPoint],
    pls: GeoPoint | None,
    missing_minutes: int,
    transcript: str,
) -> list[LikelyLocation]:
    if len(trail) < 2:
        return []

    minutes = max(1, missing_minutes)
    lowered = transcript.lower()
    pace = INJURED_PACE_MPS if any(cue in lowered for cue in _INJURY_CUES) else ON_TRAIL_PACE_MPS
    max_meters = minutes * 60 * pace
    along = _along_trail_meters(trail)
    snap = _nearest_index(trail, pls) if pls is not None else 0
    pls_along = along[snap]
    left_trail = any(cue in lowered for cue in _LEFT_TRAIL_CUES)

    candidates: list[LikelyLocation] = []
    for index, point in enumerate(trail):
        distance = abs(along[index] - pls_along)
        if distance > max_meters:
            continue
        extras: list[str] = []
        bonus = 0.0
        if _is_junction(trail, index):
            bonus += 0.15
            extras.append("switchback / decision point")
        if index in (0, len(trail) - 1):
            bonus += 0.08
            extras.append("trail end")
        if any(cue in lowered for cue in _SWITCHBACK_CUES) and _is_junction(trail, index):
            bonus += 0.2
            extras.append("switchback mentioned")
        if any(cue in lowered for cue in _VIEWPOINT_CUES) and index >= int(len(trail) * 0.55):
            bonus += 0.2
            extras.append("viewpoint cue")
        if any(cue in lowered for cue in _Y_CUES) and index >= int(len(trail) * 0.7):
            bonus += 0.2
            extras.append("Y landmark")
        if any(cue in lowered for cue in _UPHILL_CUES) and along[index] > pls_along:
            bonus += 0.1
            extras.append("uphill of PLS")
        decay = math.exp(-0.5 * ((distance - MEDIAN_DISTANCE_M) / SPREAD_M) ** 2)
        raw = decay + bonus
        reason = _reason(distance, extras)
        candidates.append(
            LikelyLocation(
                point=point,
                score=raw,
                reason=reason,
                distance_from_pls_meters=distance,
            )
        )
        if _is_junction(trail, index):
            offset = _offset_perpendicular(trail, index, OFF_TRAIL_STEP_M)
            leave_bonus = 0.12 if left_trail else 0.06
            candidates.append(
                LikelyLocation(
                    point=offset,
                    score=raw + leave_bonus,
                    reason=f"{reason}; possible leave at decision point",
                    distance_from_pls_meters=distance,
                )
            )

    kept = _suppress_near_duplicates(candidates)
    if not kept:
        return []
    peak = max(location.score for location in kept)
    if peak <= 0:
        return []
    ranked = [
        LikelyLocation(
            point=location.point,
            score=round(location.score / peak, 3),
            reason=location.reason,
            distance_from_pls_meters=round(location.distance_from_pls_meters, 1),
        )
        for location in kept
    ]
    ranked.sort(key=lambda location: location.score, reverse=True)
    return ranked[:TOP_N]


def _reason(distance: float, extras: list[str]) -> str:
    base = f"Along-trail, {distance / 1000:.1f} km from PLS"
    if extras:
        return f"{base} — {', '.join(extras)}"
    return f"{base} — time-reachable"


def _along_trail_meters(trail: Sequence[GeoPoint]) -> list[float]:
    distances = [0.0]
    for index in range(1, len(trail)):
        distances.append(distances[-1] + haversine_m(trail[index - 1], trail[index]))
    return distances


def _nearest_index(trail: Sequence[GeoPoint], point: GeoPoint) -> int:
    return min(range(len(trail)), key=lambda index: haversine_m(trail[index], point))


def _is_junction(trail: Sequence[GeoPoint], index: int) -> bool:
    if index <= 0 or index >= len(trail) - 1:
        return False
    incoming = _bearing_degrees(trail[index - 1], trail[index])
    outgoing = _bearing_degrees(trail[index], trail[index + 1])
    delta = abs((outgoing - incoming + 180) % 360 - 180)
    return delta >= TURN_DEGREES


def _bearing_degrees(start: GeoPoint, end: GeoPoint) -> float:
    lat1 = math.radians(start.lat)
    lat2 = math.radians(end.lat)
    dlng = math.radians(end.lng - start.lng)
    x = math.sin(dlng) * math.cos(lat2)
    y = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(dlng)
    return (math.degrees(math.atan2(x, y)) + 360) % 360


def _offset_perpendicular(trail: Sequence[GeoPoint], index: int, meters: float) -> GeoPoint:
    if index <= 0:
        start, end = trail[0], trail[1]
    else:
        start, end = trail[index - 1], trail[index]
    bearing = math.radians((_bearing_degrees(start, end) + 90) % 360)
    north = meters * math.cos(bearing)
    east = meters * math.sin(bearing)
    return GeoPoint(
        lat=trail[index].lat + north / 111_320.0,
        lng=trail[index].lng + east / (111_320.0 * math.cos(math.radians(trail[index].lat))),
    )


def _suppress_near_duplicates(candidates: Sequence[LikelyLocation]) -> list[LikelyLocation]:
    ordered = sorted(candidates, key=lambda location: location.score, reverse=True)
    kept: list[LikelyLocation] = []
    for location in ordered:
        if all(haversine_m(location.point, other.point) >= MIN_SEPARATION_M for other in kept):
            kept.append(location)
    return kept
